### PR TITLE
Fixes #17438 - Ensure that isinteractive multi-emit backing fields are not public 

### LINKED ISSUE #17438: Ensure that isinteractive multi-emit backing fields are not public
dotnet interactive multiemit**+** previously made the backing fields for records, unions, E.t.c public, whereas multiemit**-** makes them internal.

Some poorly written serializers, include public fields specified with a CompilerGeneratedAttribute in the serialization stream.

This means that multiemit**-** has become a work around in scenarios that rely on fsi and serialization using these types of serializers.

This PR fixes the problem by correctly specifying fields as Private (actually internal).  As implementation details they are also internal when compiled with realsig**+**

I believe the original reason for making the backing field public was so that fsi could reliably access it from multi-emit assemblies.  Due to implementation issues there was previously a limit of 30 internal visible following assemblies.  Now that it has been corrected, all multi-emit assemblies in a session have internals visible so, making the fields internal should work correctly.

Background:
@Szer 

https://github.com/dotnet/fsharp/issues/17308#issuecomment-2243914239

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes  #17438 - Ensure that isinteractive multi-emit backing fields are not public

dotnet interactive multiemit+ previously made the backing fields for records, unions, E.t.c public, whereas multiemit-makes them internal.

Some poorly written serializers, include public fields specified with a CompilerGeneratedAttribute in the serialization stream.

This means that multiemit**-** has become a work around in scenarios that rely on fsi and serialization using these types of serializers.

This PR fixes the problem by correctly specifying fields as Private (actually internal). As implementation details they are also internal when compiled with realsig**+**

I believe the original reason for making the backing field public was so that fsi could reliably access it from multi-emit assemblies. Due to implementation issues there was previously a limit of 30 internal visible following assemblies. Now that it has been corrected, all multi-emit assemblies in a session have internals visible so, making the fields internal should work correctly.

Todo:
- [x] Implementation tests
- [x] There are some other spots that seem to special case multiemit, verify that they are no longer necessary and remove them if proved unnecessary

Background:
@Szer

https://github.com/dotnet/fsharp/issues/17308#issuecomment-2243914239

Could you verify that this fix works for your scenario ... I believe it does, but I don't really use ML.net

### CODE REVIEW

**Comment by Szer** [comment]
I'll test it in 24 hours in all scenarios where it broke for me

**Comment by KevinRansom** [comment]
@szer, thanks mate, that would help.

**Comment by Szer** [comment]
checked with snippet below

Fails on `Microsoft (R) F# Interactive version 12.8.301.0 for F# 8.0`
Works on `Microsoft (R) F# Interactive version 12.9.100.0 for F# 9.0`

@KevinRansom 

```fsharp
#r "nuget: Microsoft.ML, 3.0.1"

open Microsoft.ML
open Microsoft.ML.Data
open System.IO

[<CLIMutable>]
type SpamOrHam =
    { [<LoadColumn(0)>]
      text: string
      [<LoadColumn(1)>]
      spam: bool }

let content = """text,spam
abc,spam
123,ham
"""

let tempPath = Path.GetTempFileName()
File.WriteAllText(tempPath, content)

MLContext().Data.LoadFromTextFile<SpamOrHam>(tempPath, separatorChar = ',', hasHeader = true)
```

**Comment by Szer** [comment]
Newtonsoft.Json snippet below works everywhere 
That was fixed somewhere in .NET 6 with https://github.com/dotnet/fsharp/pull/13494
Tested it as well, just in case

```fsharp
#r "nuget: Newtonsoft.Json, 13.0.3"

open Newtonsoft.Json

type Foo = { id: int }

{ id = 1 }
|> JsonConvert.SerializeObject
|> fun x -> if x <> """{"id":1}""" then failwith x
```

**Comment by KevinRansom** [comment]
@Szer , thanks this it really helps.  
I am testing an update with a bit more cleanup code reduction which looks good.  Still have to add some additional tests.  and we will try to get this in.

Kevin

File: src/Compiler/Interactive/fsi.fs [hunk 1]
```diff
@@ -1861,7 +1861,12 @@ type internal FsiDynamicCompiler
 
         let asm =
             match opts.pdbfile, pdbBytes with
-            | (Some pdbfile), (Some pdbBytes) -> File.WriteAllBytes(pdbfile, pdbBytes)
+            | (Some pdbfile), (Some pdbBytes) ->
+                File.WriteAllBytes(pdbfile, pdbBytes)
+#if FOR_TESTING
```
Comment (psfinaki) [inline src/Compiler/Interactive/fsi.fs:1866]: Do you still want to keep it?

**Review by psfinaki — APPROVED** [review]

File: src/Compiler/Interactive/fsi.fs (hunk 1 above)
Comment (KevinRansom) [reply src/Compiler/Interactive/fsi.fs:1866]: yes

**Review by vzarytovskii — APPROVED** [review]
Is it worth adding some sort of IL test for the interactive? I don't think I've been adding support of it to the testing framework though.

**Comment by KevinRansom** [comment]
> Is it worth adding some sort of IL test for the interactive? I don't think I've been adding support of it to the testing framework though.

I would like to, but we do it in memory.  Generating IL would be a touch slower.

**Pull request merged**